src/physics/base/detail/AlongStepActionImpl.hh

```diff
@@ -0,0 +1,244 @@
+                                                       particle,
+                                                       &geo,
+                                                       phys,
+                                                       mat.make_material_view(),
+                                                       msc_step_result);
+        auto msc_result = msc_scatter(rng);
+
+        // Update full path length traveled along the step based on MSC to
+        // correctly calculate energy loss, step time, etc.
+        CELER_ASSERT(geo_step <= msc_result.step_length
+                     && msc_result.step_length <= step_limit.step);
+        step_limit.step = msc_result.step_length;
+
+        // Update direction and position
+        geo.set_dir(msc_result.direction);
+        celeritas::Real3 new_pos;
+        for (int i : celeritas::range(3))
+        {
+            new_pos[i] = geo.pos()[i] + msc_result.displacement[i];
+        }
+        geo.move_internal(new_pos);
+    }
+    else
+    {
+        // Step might have changed due to geometry boundary
+        step_limit.step = geo_step;
+    }
+
+    // TODO: update track's lab-frame time here
+
+    using Energy = ParticleTrackView::Energy;
+    Energy eloss
+        = calc_energy_loss(cutoffs, mat, particle, phys, step_limit.step, rng);
+    if (eloss == particle.energy())
+    {
+        // Particle lost all energy over the step
+        if (CELER_UNLIKELY(step_limit.action != phys.scalars().range_action()
+                           && !use_msc))
+        {
+            // Particle *was* not range-limited. This means it physically moved
+            // too far over the step, since the range is supposed to be the
+            // integral of the inverse energy loss rate. Back particle slightly
+            // away from boundary to avoid on-surface initialization/direction
+            // change.
+            // NOTE: this treatment cannot be used if MSC or magnetic fields
+            // are active, because the direction has changed since the start of
+            // the step!
+            real_type backward_bump = real_type(-1e-5) * step_limit.step;
+            // Force the step limiter to be "range"
+            step_limit.action = phys.scalars().range_action();
+            step_limit.step += backward_bump;
+
+            celeritas::Real3 pos = geo.pos();
+            axpy(backward_bump, geo.dir(), &pos);
+            geo.move_internal(pos);
+        }
+
+        if (!phys.has_at_rest())
+        {
+            // Immediately kill stopped particles with no at rest processes
+            sim.status(TrackStatus::killed);
+        }
+        else
+        {
+            // Particle slowed down to zero: force a discrete interaction
+            // TODO: this is the same as the "particle.is_stopped()" condition
+            // at the end of calc_physics_step_limit: should we eliminate there
+            // or here or neither?
```

Comment:
What might we eliminate there? Setting `limit.step = 0`? Hmm trying to think it through.. if we have e.g. a positron that starts the step with zero energy, will its step limit from the previous step (step = 0, action = discrete_action) be preserved?

Reply:
I think this comment might be stale too, but I think it needs to be replaced with
```c++
step_limit.action = phys.scalars().discrete_action();
```
to avoid waiting till the beginning of the next step? Or will that be wasteful because the cross sections (for a positron with nonzero energy) are invalid at that point?

Reply:
I think that's right. It will allow it to undergo a discrete interaction, although not necessarily _force_ an interaction since the process is sampled based on pre-step macro xs, so e.g. a stopped positron might sample bremsstrahlung and then reject the interaction.